### PR TITLE
Avatars: Fetch avatars for websites with taken usernames

### LINKED ISSUE #35: Fetch avatar URLs from each website
Most of the websites have some sort of avatar which can be obtained and included in the results for 'Taken' names.

This will help to see whether a username has a consistent avatar across various platforms, which can be used as a proxy to guess that they are taken by the same person.

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Adds `get_avatar` to fetch avatars from all websites. Currently works for everything except Behance. Only works for Facebook if `profile == visible`(as expected). The function expects the user's page to exist, but handles most errors gracefully even if it doesn't(`['pinterest', 'gitlab']` are offenders of this).

Adds appropriate tests, maybe they should be mixed with test_username_check, but I think the repo could use some more breaking down into parts, perhaps this could be a start.

Closes #35

### CODE REVIEW

File: username_api.py [hunk 1]
```diff
@@ -1,22 +1,66 @@
...
+from bs4 import BeautifulSoup
 
+import os
 import sys
 import re
 import yaml
 
 app = Flask(__name__)
-cors = CORS(app) 
+cors = CORS(app)
```
Comment (andrewda) [inline username_api.py]: Unrelated change

File: username_api.py [hunk 2]
```diff
@@ -110,4 +158,4 @@ def check(website, username):
 	return jsonify(check_username(website, username))
 
 if __name__ == '__main__':
-	app.run(host='0.0.0.0', port=8521)
\ No newline at end of file
+	app.run(host='0.0.0.0', port=8521)
```
Comment (andrewda) [inline username_api.py]: Unrelated change, #41 will take care of this.

Comment (nalinbhardwaj) [reply username_api.py]: Well, none of these are conflicting/breaking changes, I guess it's fine for them to be here as well.

Comment (andrewda) [reply username_api.py]: If you're gonna keep them here, you should do them in a new commit so there's not a bunch of random changes in a commit that adds a new feature 👍

Comment (nalinbhardwaj) [reply username_api.py]: Lol, it's just a newline at EOF, nonetheless, I've undone those changes.

Comment (andrewda) [reply username_api.py]: Yea, sorry for being nitpicky, it's just generally distracting from the purpose of the PR, and makes the `git blame` look weird.

File: tests/test_data.yml [hunk 3]
```diff
@@ -1,5 +1,8 @@
 ---
 facebook:
+  avatar_usernames:
```
Comment (jayvdb) [inline tests/test_data.yml]: why do we need different usernames for facebook?

File: websites.yml [hunk 4]
```diff
@@ -46,3 +46,40 @@ username_patterns:
...
+    url: null
+  github:
+    type: meta
+    property: og:image
+    url: null
+  tumblr:
+    type: meta
+    property: og:image
+    url: null
+  behance:
```
Comment (jayvdb) [inline websites.yml]: Use `behance: false` , and order the keys alphabetically, so behance is at the top of the `avatar` group.

File: websites.yml [hunk 5]
```diff
@@ -46,3 +46,40 @@ username_patterns:
     invalid_patterns:
       - "\\.(com|net)"
       - "(\\.)\\1{1,}"  # consecutive '.' not allowed
+avatar:
+  pinterest:
+    type: regex
+    property: image_xlarge_url
+    url: null
```
Comment (jayvdb) [inline websites.yml]: I would prefer to not have empty keys in yaml unless strictly necessary.

File: websites.yml [hunk 6]
```diff
@@ -46,3 +46,40 @@ username_patterns:
     invalid_patterns:
       - "\\.(com|net)"
       - "(\\.)\\1{1,}"  # consecutive '.' not allowed
+avatar:
+  pinterest:
+    type: regex
```
Comment (jayvdb) [inline websites.yml]: `type` isnt needed.  it can be inferred. we need `property:` , `url:` , otherwise the entries can be short and clear `gitlab: opengraph` , as that is the most common.

File: tests/test_get_avatar.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import logging
+import username_api
+import os.path
+
+data = yaml.load(open(os.path.join('tests', 'test_data.yml')))
+websites = yaml.load(open('websites.yml'))
+
+class TestGet_avatar(object):
+
+	@pytest.mark.parametrize('user', data['github']['taken_usernames'])
```
Comment (manu-chroma) [inline tests/test_get_avatar.py]: Can we have better test parameterisation in this PR? (specifically for the newly added tests) There is lot of duplicated logic already in tests.

File: username_api.py [hunk 8]
```diff
@@ -12,11 +14,53 @@
...
+	if data['type'] == 'direct':
+		if 'image' not in response.headers.get('content-type'):
+			return None
+		return url
+	else:
+		response = r.get(url)
+		soup = BeautifulSoup(response.text, 'html.parser')
+		if data['type'] == 'meta':
+			# Look in metadata for `property` attributed image.
+			result = [item.attrs['content'] for item in soup('meta') if item.has_attr('property') and item.attrs['property'].lower()==data['property']]
```
Comment (manu-chroma) [inline username_api.py]: split this line

File: username_api.py [hunk 9]
```diff
@@ -12,11 +14,53 @@
...
+			if not result or not result[0]:
+				return None
+			result = result[0]
+		elif data['type'] == 'regex':
+			# Searches for "`property`": "`link`"
+			regex = re.compile('[\'\"]' + re.escape(data['property']) + '[\'\"]:(\s)?[\'\"](?P<link>[^\s]+)[\'\"]')
```
Comment (manu-chroma) [inline username_api.py]: split

File: tests/test_data.yml (hunk 3 above)
Comment (nalinbhardwaj) [reply tests/test_data.yml]: As noted in the PR description, profiles on Facebook can be “hidden”, therefore having no avatar to display. These are public profile links.

Comment (jayvdb) [reply tests/test_data.yml]: ok

File: tests/test_get_avatar.py (hunk 7 above)
Comment (nalinbhardwaj) [reply tests/test_get_avatar.py]: Okay, I will look into parameterisation. I would assume it's okay to use the same methods/parameterisation in the username_check tests. Is that fine?(Of course, I'll do those in a separate PR later)

Comment (jayvdb) [reply tests/test_get_avatar.py]: If more parameterisation is not easy, it can be deferred to GCI task https://github.com/manu-chroma/username-availability-checker/issues/36

File: username_api.py [hunk 10]
```diff
@@ -12,11 +14,56 @@
 
 patterns = yaml.load(open('websites.yml'))
 
-def check_username(website, username):
-	url = patterns['urls'].get(website, 'https://{w}.com/{u}').format(
+def get_url(website, username):
```
Comment (jayvdb) [inline username_api.py]: `get_profile_url` ?

File: websites.yml [hunk 11]
```diff
@@ -46,3 +46,21 @@ username_patterns:
     invalid_patterns:
       - "\\.(com|net)"
       - "(\\.)\\1{1,}"  # consecutive '.' not allowed
+avatar:
+  pinterest:
+    property: image_xlarge_url
+  gitlab:
+    property: og:image
```
Comment (jayvdb) [inline websites.yml]: This isnt a property.  This is a standard.  It is *the* standard.  If you dont like `gitlab: true` to reflect using OpenGraph, then please use

Comment (nalinbhardwaj) [reply websites.yml]: @jayvdb : Not sure if I understand what you mean correctly but this is for future compatibility, if there is a website added in future which has `twitter:image`( see [this](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/overview/summary) ) but not `og:image`, all that needs to be done is just write `twitter:image` here(instead of `og:image`) for it to work.
This also allows for supporting arbitrary meta tags for the image.

Comment (nalinbhardwaj) [reply websites.yml]: Maybe you mean this should be renamed to `meta-name` or something?
Or that this is an unnecessary feature? Let me know if it's the second case, I'll remove it.

Comment (jayvdb) [reply websites.yml]: What I want is `gitlab: true`.  OpenGraph is the default.  Nothing else is needed.  Any other alternative is an override.  And that can be designed later.

Comment (nalinbhardwaj) [reply websites.yml]: Ok, so I'll shift this to
```
gitlab:
  opengraph: true
```
so we can still have url overrides easily.

File: tests/test_get_avatar.py (hunk 7 above)
Comment (nalinbhardwaj) [reply tests/test_get_avatar.py]: I'm going to shift us from pytest to nosetests as part of #36 I think. It has a pretty powerful library [parameterized](https://github.com/wolever/parameterized), which although supports py.test, isn't as well supported(not as pretty and I see quite a few bug reports) (see [FAQ](https://github.com/wolever/parameterized#faq) for more details).

This is probably a large change(to builds etc.) , so I'd want it to be in it's own PR, but let me know whatever you decide.

File: websites.yml (hunk 11 above)
Comment (jayvdb) [reply websites.yml]: `gitlab: true` just coexists fine with
```
facebook:
  url: ...
```
This is YAML.
You are using YAML to store data in a JSON-like structure.  That is not YAML.

File: tests/test_get_avatar.py (hunk 7 above)
Comment (jayvdb) [reply tests/test_get_avatar.py]: nosetests is dead, and nose2 is stillborn.
It works well in pytest.  The caveat about the generated method names is of little consequence as usually test method names are hidden, and `name_func` can be used to create new names that are meaningful.

Comment (nalinbhardwaj) [reply tests/test_get_avatar.py]: Are we okay dealing [parameterized/#34](https://github.com/wolever/parameterized/issues/34) then?
Or can we just shift to unittest(since we probably won't need pytest with that)?

Comment (jayvdb) [reply tests/test_get_avatar.py]: Not our problem.  pytest devs are not going to delete remove support for that until their main plugins are fixed.  That is just an early warning system, and the alarm has been heard correctly.

File: username_api.py [hunk 12]
```diff
@@ -12,11 +14,55 @@
...
+		result = result[0]
+	elif 'property' in data:
+		# Searches for "`property`": "`link`"
+		regex = re.compile('[\'\"]' + re.escape(data['property']) +
+						   '[\'\"]:(\s)?[\'\"](?P<link>[^\s]+)[\'\"]')
+		result = re.search(regex, response.text)
+		if not result:
+			return None
+		result = result.group('link')
+	elif 'image' in response.headers.get('content-type'):
```
Comment (jayvdb) [inline username_api.py]: `response.headers.get('content-type')` can return None.

`...startswith('image/')`

File: username_api.py [hunk 13]
```diff
@@ -12,11 +14,55 @@
...
+
+	url = get_profile_url(website, username)
+	if 'url' in data:
+		url = data['url'].format(u=username)
+
+	response = r.get(url)
+	if response.status_code == 404:
+		return None
+
+	soup = BeautifulSoup(response.text, 'html.parser')
```
Comment (jayvdb) [inline username_api.py]: this should be inside the relevant branch of the `if` ; it is an expensive object to create

File: tests/test_get_avatar.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+		parameterized.to_safe_name('_'.join(str(x) for x in param.args)),
+	)
+
+
+class TestGet_avatar(object):
+
+	@parameterized.expand(load_test_cases('with'),
+						  testcase_func_name=custom_name_func)
+	def test_with_avatar(self, website, user):
+		if website == 'behance':
```
Comment (jayvdb) [inline tests/test_get_avatar.py]: this belongs in `load_test_cases`

File: tests/test_get_avatar.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+			pytest.skip("behance doesn't work")
+		link = username_api.check_username(website, user)['avatar']
+		response = r.get(link)
+		assert('image' in response.headers.get('content-type') or
```
Comment (jayvdb) [inline tests/test_get_avatar.py]: `startswith(`image/`)`

File: tests/test_get_avatar.py (hunk 14 above)
Comment (nalinbhardwaj) [reply tests/test_get_avatar.py]: So we shouldn't even load those test cases? I'd rather keep a "skipped" notification no?

Comment (jayvdb) [reply tests/test_get_avatar.py]: you can generate a skipped test in `load_test_cases`

File: username_api.py [hunk 16]
```diff
@@ -12,11 +14,56 @@
...
 patterns = yaml.load(open('websites.yml'))
 
-def check_username(website, username):
-	url = patterns['urls'].get(website, 'https://{w}.com/{u}').format(
+def get_profile_url(website, username):
+	return patterns['urls'].get(website, 'https://{w}.com/{u}').format(
 		w=website,
-		u=username
-	)
+		u=username)
```
Comment (jayvdb) [inline username_api.py]: unnecessary change here

File: username_api.py [hunk 17]
```diff
@@ -12,11 +14,56 @@
...
+		result = result[0]
+	elif 'property' in data:
+		# Searches for "`property`": "`link`"
+		regex = re.compile('[\'\"]' + re.escape(data['property']) +
+						   '[\'\"]:(\s)?[\'\"](?P<link>[^\s]+)[\'\"]')
+		result = re.search(regex, response.text)
+		if not result:
+			return None
+		result = result.group('link')
+	elif (response.headers.get('content-type') and
```
Comment (jayvdb) [inline username_api.py]: I would use `response.headers.get('content-type', '').startswith('image/')` to make it more readable and only perform slightly worse in the very unlikely scenario that one of these sites sends a response without a content-type

File: username_api.py (hunk 16 above)
Comment (nalinbhardwaj) [reply username_api.py]: This is actually not a very good diff comparison, I have a fresh function in those lines, for some reason because of the repetition from the old one, git decides it's a modification of that. I'll change it nonetheless.

File: username_api.py (hunk 17 above)
Comment (jayvdb) [reply username_api.py]: actually, that would still perform better, because every `.` is a perf hit using `getattribute` so your `and` here is doubling the number of times `response.headers.get('content-type')` occurs.

File: username_api.py (hunk 16 above)
Comment (jayvdb) [reply username_api.py]: it is a very good comparison. :P
Splitting a function like this is very common.
not changing the lines in the middle helps code review and understanding what is changing and what isnt.

File: username_api.py (hunk 17 above)
Comment (nalinbhardwaj) [reply username_api.py]: Well, optimally, by locality of reference, the perf hit would be negligible, but we're dealing with python so performance^TM 🙃

File: websites.yml [hunk 18]
```diff
@@ -46,3 +46,15 @@ username_patterns:
     invalid_patterns:
       - "\\.(com|net)"
       - "(\\.)\\1{1,}"  # consecutive '.' not allowed
+avatar:
+  pinterest:
+    property: image_xlarge_url
```
Comment (manu-chroma) [inline websites.yml]: is ``property`` a good term for this?

Comment (nalinbhardwaj) [reply websites.yml]: how about `key` if you don't like it?

Comment (manu-chroma) [reply websites.yml]: i think ``key`` would be better since we're picking from ``json`` data. ``property`` seems vague here

Comment (nalinbhardwaj) [reply websites.yml]: done.

File: username_api.py [hunk 19]
```diff
@@ -12,12 +14,57 @@
...
+	if response.status_code == 404:
+		return None
+
+	if data == 'opengraph':
+		# Look in metadata for image.
+		soup = BeautifulSoup(response.text, 'html.parser')
+		result = [item.attrs['content'] for item in soup('meta')
+				  if item.has_attr('property') and
+				  item.attrs['property'].lower() == 'og:image']
+		if not result or not result[0]:
```
Comment (manu-chroma) [inline username_api.py]: why is check for both ``result`` and ``result[0]`` needed here?

File: username_api.py [hunk 20]
```diff
@@ -12,12 +14,57 @@
...
+		result = result[0]
+	elif 'key' in data:
+		# Searches for "`key`": "`link`"
+		regex = re.compile('[\'\"]' + re.escape(data['key']) +
+						   '[\'\"]:(\s)?[\'\"](?P<link>[^\s]+)[\'\"]')
+		result = re.search(regex, response.text)
+		if not result:
+			return None
+		result = result.group('link')
+	elif response.headers.get('content-type', '').startswith('image/'):
```
Comment (manu-chroma) [inline username_api.py]: this one applies for twitter?

Comment (nalinbhardwaj) [reply username_api.py]: yes.

File: username_api.py (hunk 19 above)
Comment (nalinbhardwaj) [reply username_api.py]: Yes, some website returned an empty string for some reason. This is for that. (I can't remember which one, I think it was fb)

File: tests/test_get_avatar.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+def custom_name_func(testcase_func, param_num, param):
+	return '%s_%s' % (
+		testcase_func.__name__,
+		parameterized.to_safe_name('_'.join(str(x) for x in param.args)),
+	)
+
+
+class TestGet_avatar(object):
+
+	@parameterized.expand(load_test_cases('with'),
```
Comment (manu-chroma) [inline tests/test_get_avatar.py]: can we do ``with_avatar`` and ``without_avatar`` here? Other than that this PR is good to go.

File: websites.yml [hunk 22]
```diff
@@ -46,3 +46,15 @@ username_patterns:
     invalid_patterns:
       - "\\.(com|net)"
       - "(\\.)\\1{1,}"  # consecutive '.' not allowed
+avatar:
+  pinterest:
```
Comment (jayvdb) [inline websites.yml]: alpha sort the websites?
(it isnt done elsewhere, but that is a cleanup for later.)

**Comment by manu-chroma** [comment]
good job @nalinbhardwaj :smile:
